dace/transformation/dataflow/map_distribution.py

```diff
@@ -0,0 +1,940 @@
+        # See, for example, samples/simple/mandelbrot.py
+        for node in subgraph_from_maps(sdfg, graph, [map_entry]):
+            if isinstance(node, dace.nodes.CodeNode):
+                for p in params:
+                    if str(p) in node.free_symbols:
+                        return False
+
+        inputs = dict()
+        for _, _, _, _, m in graph.out_edges(map_entry):
+            if not m.data:
+                continue
+            desc = sdfg.arrays[m.data]
+            if desc not in inputs.keys():
+                inputs[desc] = []
+            inputs[desc].append(m.subset)
+        
+        for desc, accesses in inputs.items():
+            if isinstance(desc, dace.data.Scalar):
+                continue
+            elif isinstance(desc, (dace.data.Array, dace.data.View)):
+                if list(desc.shape) == [1]:
+                    continue
+                for a in accesses:
+                    if a.num_elements() != 1:
+                        return False
+                    indices = a.min_element()
+                    unmatched_indices = set(params)
+                    for idx in indices:
+                        if idx in unmatched_indices:
+                            unmatched_indices.remove(idx)
+                    if len(unmatched_indices) > 0:
+                        return False
+            else:
+                return False
+
+        outputs = dict()
+        for _, _, _, _, m in graph.in_edges(map_exit):
+            if m.wcr:
+                return False
+            desc = sdfg.arrays[m.data]
+            if desc not in outputs.keys():
+                outputs[desc] = []
+            outputs[desc].append(m.subset)
+        
+        for desc, accesses in outputs.items():
+            if isinstance(desc, (dace.data.Array, dace.data.View)):
+                for a in accesses:
+                    if a.num_elements() != 1:
+                        return False
+                    indices = a.min_element()
+                    unmatched_indices = set(params)
+                    for idx in indices:
+                        if idx in unmatched_indices:
+                            unmatched_indices.remove(idx)
+                    if len(unmatched_indices) > 0:
+                        return False
+            else:
+                return False
+
+        return True
+
+    @staticmethod
+    def match_to_str(graph: dace.SDFGState, candidate: Dict[pm.PatternNode,
+                                                            int]) -> str:
+        map_entry = graph.node(candidate[ElementWiseArrayOperation._map_entry])
+        return map_entry.map.label + ': ' + str(map_entry.map.params)
+
+
+    def apply(self, sdfg: dace.SDFG):
+        graph = sdfg.nodes()[self.state_id]
+        map_entry = graph.nodes()[self.subgraph[self._map_entry]]
+        map_exit = graph.exit_node(map_entry)
+
+        sz = dace.symbol('commsize', dtype=dace.int32)
+
+        def _prod(sequence):
```

Comment:
can be imported from `dace.data` instead

Reply:
I agree that we shouldn't define `prod` again and again, but we should import from some general utility file. Importing from `dace.data` is very strange.